### PR TITLE
feat: add option to specify semantic version category

### LINKED ISSUE #459: Filter PR to auto merge with semantic version
**What do you want to achieve?**

GitHub is encouraging people to move to [Github-native dependabot](https://github.blog/2020-06-01-keep-all-your-packages-up-to-date-with-dependabot/#moving-forward-from-dependabot-com-and-dependabot-preview). As you know they removed auto-merge function.

With dependabot preview we are using NPM semantic versioning so that it only auto-merges minors and patches version, leaving major versions alone. It leaves us the opportunity to review (and learn) about major changes in the library (even if they don't break the tests) while making it as low maintenance as possible for other upgrades. We can arrive in the morning and all the _safe/easy_ upgrades are done.

This is our current setting in dependabot-preview:

<kbd>

</kbd>

I think it is not possible to do this right now with _merge-me-action_ (and maybe it is just too hard to implement).

**What is the current way of working?**

Using the tandem Github-native dependabot we are either forced to auto-merge everything with _merge-me-action_ or manually merge all upgrades PR.

**How much does it hurt?**
It is unsustainable for us, so, for now, we plan to stay as long as possible on dependabot-preview.

### PR DESCRIPTION
**Pull request by dunyakirkali**
This Pull Request fulfills the following requirements:

- [x] The commit message follows our guidelines.
- [x] Tests for the changes have been added if needed.
- [x] Does not affect documentation or it has been added or updated.

Resolves #459

### CODE REVIEW

File: src/utilities/prTitleParser.spec.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import { parsePRTitle } from './prTitleParser';
+
+var title: string = '';
```
Comment (dunyakirkali) [inline src/utilities/prTitleParser.spec.ts:3]: @vdvleon wat is a better way of doing this?

File: src/utilities/inputParsers.ts [hunk 2]
```diff
@@ -21,3 +27,17 @@ export const parseInputMergeMethod = (): AllowedMergeMethods => {
...
+
+  if (input.length === 0 || AllowedMergeCategories[input] === undefined) {
+    logWarning(
+      'MERGE_CATEGORY value input is ignored because its malformed, defaulting to MAJOR.',
+    );
+
+    return AllowedMergeCategories.MAJOR;
+  }
+
+  return AllowedMergeCategories[input];
```
Comment (vdvleon) [inline src/utilities/inputParsers.ts:42]: Do you know for sure here that input is always one of the categories?

File: src/eventHandlers/pullRequest/index.ts [hunk 3]
```diff
@@ -89,6 +90,15 @@ export const pullRequestHandle = async (
       )}.`,
     );
 
+    // Check merge category
+    const category = getInput('MERGE_CATEGORY');
+    if (!parsePRTitle(pullRequest.title, category)) {
```
Comment (dunyakirkali) [inline src/eventHandlers/pullRequest/index.ts:95]: Does this mean

```suggestion
    if ( !parsePRTitle(pullRequest.title, category)) {
```

File: src/utilities/prTitleParser.spec.ts (hunk 1 above)
Comment (vdvleon) [inline src/utilities/prTitleParser.spec.ts:3]: We don't use global vars (and especially not `var`).

File: src/utilities/prTitleParser.spec.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+describe('parsePRTitle', (): void => {
+  describe('given containing MAJOR bump', (): void => {
+    beforeEach((): void => {
+      title = 'bump @types/jest from 26.0.12 to 27.0.13';
+    });
```
Comment (vdvleon) [inline src/utilities/prTitleParser.spec.ts:9]: I would just write this:
```suggestion
    const title = 'bump @types/jest from 26.0.12 to 27.0.13';
```

Comment (vdvleon) [reply src/utilities/prTitleParser.spec.ts:9]: Should the title have a 'type' ? Like `chore: bump ...`

File: src/utilities/prTitleParser.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
+// TODO: (dunyakirkali) Better naming
+export const parsePRTitle = (title: string, category: string): boolean => {
```
Comment (vdvleon) [inline src/utilities/prTitleParser.ts:2]: Would it be possible to use the enum type here?

File: src/utilities/prTitleParser.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  const match = semVerTitleRegExp.exec(title);
+
+  if (match) {
+    const semVerRegExp = /^(?<major>0|[1-9]\d*)\.(?<minor>0|[1-9]\d*)\.(?<patch>0|[1-9]\d*)$/u;
+
+    // TODO: (dunyakirkali) Handle optionals
+    const from = match.groups?.from!;
+    const to = match.groups?.to!;
+
+    // TODO: (dunyakirkali) Handle optionals
```
Comment (vdvleon) [inline src/utilities/prTitleParser.ts:17]: TODOs should have a deadline.

**Review by vdvleon — CHANGES_REQUESTED** [review]
Logic looks good, but I gave some (linting related) comments.

File: src/utilities/inputParsers.spec.ts [hunk 7]
```diff
@@ -32,3 +35,32 @@ describe('parseInputMergeMethod', (): void => {
...
+
+  it('returns default merge category if merge category is not allowed', (): void => {
+    expect.assertions(1);
+
+    getInputSpy.mockReturnValueOnce('OTHER');
+
+    expect(parseInputMergeCategory()).toStrictEqual('MAJOR');
+  });
+
+  it('returns undefined if merge category is not provided', (): void => {
```
[the remark this answers is not in the document]
Comment (dunyakirkali) [reply src/utilities/inputParsers.spec.ts:59]: good catch

File: src/utilities/prTitleParser.ts (hunk 6 above)
Comment (vdvleon) [reply src/utilities/prTitleParser.ts:17]: Example: `TODO (dunyakirkali) [2020-10-01]: Handle optionals.`

Comment (dunyakirkali) [reply src/utilities/prTitleParser.ts:17]: so what is the convention in TS to handle these optional values?

Do I have to pepper the code with `=== undefined` ?

Comment (vdvleon) [reply src/utilities/prTitleParser.ts:17]: Yeah, or you could default to some value. E.g. `const to = match.groups?.to ?? '';`

File: src/utilities/inputParsers.ts (hunk 2 above)
Comment (dunyakirkali) [reply src/utilities/inputParsers.ts:42]: yeah. The github action handles that bit.

File: src/utilities/prTitleParser.ts [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    const semVerRegExp = /^(?<major>0|[1-9]\d*)\.(?<minor>0|[1-9]\d*)\.(?<patch>0|[1-9]\d*)$/u;
+
+    const matchGroups = match.groups;
+
+    if (matchGroups === undefined || matchGroups.to === undefined || matchGroups.from === undefined) {
+      return true;
+    }
+
+    const from = matchGroups!.from!;
+    const to = matchGroups!.to!;
```
Comment (dunyakirkali) [inline src/utilities/prTitleParser.ts:20]: @vdvleon something like this? It just looks so ugly

Comment (vdvleon) [reply src/utilities/prTitleParser.ts:20]: Nope. You can never use `!`. I would just write:

```typescript
const from = matchGroups?.from;
const to = matchGroups?.to;

if (from === undefined || to === undefined) {
  return true;
}
```

After this if-statement, TypeScript understands that `to` and `from` are not `undefined`, so in this case a string.

**Review by vdvleon — APPROVED** [review]
👍

**Comment by dunyakirkali** [comment]
@merlinnot Linting is failing because of an expired todo. How should I go about that?

File: src/eventHandlers/pullRequest/index.ts [hunk 9]
```diff
@@ -89,6 +91,19 @@ export const pullRequestHandle = async (
...
+      checkPullRequestTitleForMergeCategory(
+        pullRequest.title,
+        getInput('MERGE_CATEGORY'),
+      ) === false
+    ) {
+      logInfo(
+        'Skipping auto-merge since the upgrade does not match merge category',
+      );
+
+      return;
```
Comment (dunyakirkali) [inline src/eventHandlers/pullRequest/index.ts:104]: @merlinnot I couldn't find a good way of mocking the PR title.

Do you have some pointers for me on how I can go about testing this bit?

**Comment by merlinnot** [comment]
> @merlinnot Linting is failing because of an expired todo. How should I go about that?

Bump it and ping a person responsible 😉

**Comment by dunyakirkali** [comment]
> > @merlinnot Linting is failing because of an expired todo. How should I go about that?
> 
> 
> 
> Bump it and ping a person responsible 😉 

Should I do that in this PR (I think not).

Isn't it nicer if that person (u in this case ;) ) makes a pr to master and I rebase my branch to master?

**Comment by merlinnot** [comment]
Absolutely: https://github.com/ridedott/merge-me-action/pull/540.

File: README.md [hunk 10]
```diff
@@ -142,6 +142,23 @@ steps:
       MERGE_METHOD: MERGE
 ```
 
+### Specify the semantic versioning category to use
+
+By default, this GitHub Action assumes merge category is `MAJOR`. This feature
+looks at the title of the PR to figure out what is being updated. We expect a
```
Comment (merlinnot) [inline README.md:148]: Let's use third person passive voice, I don't know who is "we" here 👨‍🎓

File: README.md [hunk 11]
```diff
@@ -142,6 +142,23 @@ steps:
...
+dependabot style title such as: `bump @types/jest from 26.0.12 to 27.0.13`. You
```
Comment (merlinnot) [inline README.md:149]: Well, no one expects such PRs, this Action is not only for Dependabot, up until now I don't think it favorized it in any way. Are we going to support Renovate, Gardener and other tools as well? Would you consider changing renaming `MERGE_CATEGORY` to `PRESET` and values to `DEPENDABOT_MAJOR`, `DEPENDABOT_MINOR`, ...? That way it would be communicated better that it's open to adding more presets, e.g. presets for languages that don't use semantic versioning at all.

Comment (dunyakirkali) [reply README.md]: Initially I'm not planning on supporting Renovate, Gardener or others. If such a request would come in, then we can extend this logic to support those platforms as well.

Comment (dunyakirkali) [reply README.md]: I renamed `MERGE_CATEGORY` to `PRESET`

Comment (dunyakirkali) [reply README.md]: @merlinnot maybe we could use something like `SEMANTIC_MAJOR` instead of `DEPENDABOT_MAJOR` WDYT?

Since some (not all) of the other tools will be using semantic versioning

File: src/eventHandlers/pullRequest/test.spec.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
+/**
+ * @webhook-pragma pull_request_for_major_bump
```
Comment (dunyakirkali) [inline src/eventHandlers/pullRequest/test.spec.ts:2]: @viestat you're suggestion of creating a new test file did work although I'm not quite sure what I should call this file. Any suggestions?

File: README.md (hunk 11 above)
Comment (merlinnot) [reply README.md]: Well, I just want to avoid any breaking changes in the future. It seems to me that the feature we built here is very specific to Dependabot, especially that it relies on a very specific PR title that Dependabot creates. By naming it `DEPENDABOT_MAJOR` we can manage expectations a little better. If we call it `SEMANTIC_MINOR`, one can also expect that it will work with Renovate (merge only patch and minor from it).

File: README.md [hunk 13]
```diff
@@ -142,6 +142,23 @@ steps:
       MERGE_METHOD: MERGE
 ```
 
+### Specify the semantic versioning category to use
+
+By default, this GitHub Action assumes merge category is `MAJOR`. This feature
+looks at the title of the PR to figure out what is being updated. A dependabot
+style title such as: `bump @types/jest from 26.0.12 to 27.0.13` is expected. You
```
Comment (merlinnot) [inline README.md:149]:
```suggestion
style title such as: `chore(deps): bump @types/jest from 26.0.12 to 27.0.13` is expected. You
```

File: README.md (hunk 11 above)
Comment (dunyakirkali) [reply README.md]: Gotcha. I'll change it 👍

File: README.md (hunk 13 above)
Comment (dunyakirkali) [reply README.md:149]: Good catch!

File: src/eventHandlers/pullRequest/index.ts (hunk 9 above)
Comment (dunyakirkali) [reply src/eventHandlers/pullRequest/index.ts]: I managed to get it tested!

File: README.md (hunk 11 above)
Comment (dunyakirkali) [reply README.md]: Done!

File: README.md [hunk 14]
```diff
@@ -144,19 +144,20 @@ steps:
 
 ### Specify the semantic versioning category to use
 
-By default, this GitHub Action assumes merge category is `MAJOR`. This feature
-looks at the title of the PR to figure out what is being updated. A dependabot
-style title such as: `chore(deps): bump @types/jest from 26.0.12 to 27.0.13` is expected. You
-can override the merge category by changing the value of `PRESET` parameter (one
-of `MAJOR`, `MINOR` or `PATCH`):
+By default, this GitHub Action assumes merge category is `DEPENDABOT_MAJOR`.
```
Comment (merlinnot) [inline README.md:147]: Well, maybe by default we should have no presets? I mean, keep the current behavior, merge everything.

Comment (dunyakirkali) [reply README.md:147]: Having `DEPENDABOT_MAJOR` means that everything will be merged (major, minor and patch updates)

File: src/eventHandlers/pullRequest/test.spec.ts (hunk 12 above)
Comment (dunyakirkali) [reply src/eventHandlers/pullRequest/test.spec.ts:2]: Renamed it to `majorBump.spec.ts` as @viestat suggested

File: README.md (hunk 14 above)
Comment (merlinnot) [reply README.md:147]: Doesn't that suggest the someone should use this action with Dependabot? I'm not sure if that's an intuitive default, I can see it causing confusion.

Comment (dunyakirkali) [reply README.md:147]: hmm, I c what you mean. I'll change as suggested

Comment (dunyakirkali) [reply README.md]: Done

File: .tool-versions [hunk 15]
```diff
@@ -0,0 +1 @@
+nodejs 10.22.0
```
Comment (merlinnot) [inline .tool-versions:1]: Did you mean to commit this file? We officially recommend nvm for Node.js, not asdf.

File: README.md [hunk 16]
```diff
@@ -142,6 +142,23 @@ steps:
       MERGE_METHOD: MERGE
 ```
 
+### Specify the semantic versioning category to use
```
Comment (merlinnot) [inline README.md:145]: I looked at the specification and didn't see any reference to "category".

I'd suggest refactoring this paragraph to:

```
### Presets

Presets enable additional functionality which can be used to better personalize
default behavior of the Merge me! Action.

Available presets are:
- `DEPENDABOT_MINOR` - Merge only minor dependency updates for pull requests created by Dependabot if the dependency version follows [Semantic Versioning v2](https://semver.org/).
- `DEPENDABOT_PATH` - Merge only patch dependency updates for pull requests created by Dependabot if the dependency version follows [Semantic Versioning v2](https://semver.org/).
```

I think we can drop `DEPENDABOT_MAJOR`, as it literally doesn't change the behavior at all?

File: action.yml [hunk 17]
```diff
@@ -18,6 +18,11 @@ inputs:
       One of: MERGE, STASH or REBASE'
     required: false
     default: SQUASH
+  PRESET:
+    description:
+      'Specify the semantic version category to automatically merge. One of:
```
Comment (merlinnot) [inline action.yml:23]:
```suggestion
      'Enable additional functionality to better personalize the behavior.`
```

File: src/utilities/prTitleParsers.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
+/* eslint-disable max-statements */
+export const checkPullRequestTitleForMergePreset = (
```
Comment (merlinnot) [inline src/utilities/prTitleParsers.ts:3]: Newline please 🙂 Yes, we should add an ESLint rule for it.

File: src/utilities/prTitleParsers.ts [hunk 19]
```diff
@@ -0,0 +1,47 @@
+/* eslint-disable @typescript-eslint/no-non-null-assertion */
```
Comment (merlinnot) [inline src/utilities/prTitleParsers.ts:1]: I don't see why we'd disable it instead of improving the code?

**Review by merlinnot — COMMENTED** [review]
The implementation look good overall 🙂 We might need to adjust it depending on how we resolve some comments about the documentation I added.

File: .tool-versions (hunk 15 above)
Comment (dunyakirkali) [reply .tool-versions]: nope 😊 
Sorry for that

File: README.md (hunk 16 above)
Comment (dunyakirkali) [reply README.md:145]: Great suggestions!

I'll update the README and remove `DEPENDABOT_MAJOR`

Comment (dunyakirkali) [reply README.md]: Done!

File: src/utilities/prTitleParsers.ts (hunk 19 above)
Comment (dunyakirkali) [reply src/utilities/prTitleParsers.ts:1]: The reason I ended up using non-null assertions was because of this: https://github.com/microsoft/TypeScript/issues/32098

So in my case I have:

```typescript
const regExp = /bump .* from (?<from>.*) to (?<to>.*)/u;
const match = regExp.exec(title);

// match can be null
if (match === null) {
  return true;
}

// but if match is not null there is no way that the `from` and `to` don't exist
// tho Typescript is not smart enough to understand that
```

But if you rather have me not have the `@typescript-eslint/no-non-null-assertion` then I can check the `undefined`ness of the `from` and `to` tho the coe would become uglier. WDYT?

Comment (merlinnot) [reply src/utilities/prTitleParsers.ts:1]: That seems like a valid reason, thanks for the explanation!

File: src/eventHandlers/pullRequest/index.spec.ts [hunk 20]
```diff
@@ -1,7 +1,6 @@
 /**
  * @webhook-pragma pull_request
  */
-
```
Comment (merlinnot) [inline src/eventHandlers/pullRequest/index.spec.ts:4]: Why the change?

File: action.yml [hunk 21]
```diff
@@ -18,6 +18,11 @@ inputs:
       One of: MERGE, STASH or REBASE'
     required: false
     default: SQUASH
+  PRESET:
+    description:
+      'Enable additional functionality to better personalize the behavior.`
+      DEPENDABOT_MINOR or DEPENDABOT_PATCH'
```
Comment (merlinnot) [inline action.yml:24]: I think we forgot to reformat this description when changing it?

File: src/utilities/inputParsers.ts [hunk 22]
```diff
@@ -21,3 +26,15 @@ export const parseInputMergeMethod = (): AllowedMergeMethods => {
 
   return AllowedMergeMethods[input];
 };
+
+export const parseInputMergePreset = (): AllowedMergePresets | undefined => {
+  const input = getInput('PRESET');
+
+  if (input.length === 0 || AllowedMergePresets[input] === undefined) {
+    logWarning('PRESET value input is ignored because its malformed.');
```
Comment (merlinnot) [inline src/utilities/inputParsers.ts:34]: Let's change to "it is" (we have "its" with no apostrophe).

File: src/eventHandlers/pullRequest/majorBump.spec.ts [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+const PULL_REQUEST_ID = 'MDExOlB1bGxSZXF1ZXN0MzE3MDI5MjU4';
+const COMMIT_HEADLINE = 'Update test';
+
+const octokit = getOctokit('SECRET_GITHUB_TOKEN');
+const getInputSpy = jest.spyOn(core, 'getInput').mockImplementation();
+
+jest.spyOn(core, 'info').mockImplementation();
+
+beforeEach((): void => {
+  getInputSpy.mockReturnValue('PATCH');
```
Comment (merlinnot) [inline src/eventHandlers/pullRequest/majorBump.spec.ts:22]: Is this up to date?

Comment (dunyakirkali) [reply src/eventHandlers/pullRequest/majorBump.spec.ts]: Corrected

File: src/eventHandlers/pullRequest/index.spec.ts (hunk 20 above)
Comment (dunyakirkali) [reply src/eventHandlers/pullRequest/index.spec.ts:4]: Oups

File: action.yml (hunk 21 above)
Comment (dunyakirkali) [reply action.yml:24]: Fixed!

File: action.yml [hunk 24]
```diff
@@ -18,6 +18,11 @@ inputs:
       One of: MERGE, STASH or REBASE'
     required: false
     default: SQUASH
+  PRESET:
+    description:
+      'Enable additional functionality to better personalize the behavior. One
+      of: DEPENDABOT_MINOR or DEPENDABOT_PATCH'
```
Comment (merlinnot) [inline action.yml:24]:
```suggestion
      of: DEPENDABOT_MINOR or DEPENDABOT_PATCH.'
```

**Review by merlinnot — APPROVED** [review]
Let's ship it! 🎉

File: README.md [hunk 25]
```diff
@@ -142,6 +142,29 @@ steps:
...
+
+Presets enable additional functionality which can be used to better personalize
+default behavior of the Merge me! Action.
+
+Available presets are:
+
+- `DEPENDABOT_MINOR` - Merge only minor and patch dependency updates for pull
+  requests created by Dependabot if the dependency version follows
+  [Semantic Versioning v2](https://semver.org/).
+- `DEPENDABOT_PATH` - Merge only patch dependency updates for pull requests
```
Comment (jraoult) [inline README.md:155]: Sorry a little late to the party, but surely here it was supposed to be `PATCH`

Comment (JrSchild) [reply README.md:155]: Thanks for pointing that out @jraoult. I opened a PR for it #545
